### PR TITLE
Coupons

### PR DESCRIPTION
**Pull request by gsnovak**
@fiestacasey

### CODE REVIEW

File: app/controllers/admin/products_controller.rb [hunk 1]
```diff
@@ -5,7 +5,7 @@ class ProductsController < AdminController
     private
 
     def product_params
-      params.require(:product).permit(:name, :available, :on_hand_count, :price, :picture)
+      params.require(:product).permit(:name, :available, :on_hand_count, :price, :picture, :amount)
```
Comment (fiestacasey) [inline app/controllers/admin/products_controller.rb]: products have amount and price? seems redundant and confusing...

File: app/models/coupon.rb [hunk 2]
```diff
@@ -3,7 +3,11 @@ class Coupon < ApplicationRecord
     active: [:deleted],
     deleted: []
    })
-
+  before_save :negate_amount
```
Comment (fiestacasey) [inline app/models/coupon.rb]: this is silly.... can't you think of a better way?

File: app/models/order_item.rb [hunk 3]
```diff
@@ -1,4 +1,7 @@
 class OrderItem < ApplicationRecord
+  validates_uniqueness_of :source_type,
+    scope: :order_id,
+    if: "[Coupon.name].include?(source_type)"
```
Comment (fiestacasey) [inline app/models/order_item.rb]: `source_type == Coupon.name` is simpler

File: app/views/admin/products/_form.html.haml [hunk 4]
```diff
@@ -5,5 +5,5 @@
     = f.hidden_field(:picture_cache)
     = f.file_field(:picture)
     = f.number_field :on_hand_count
-    = f.number_field :price, class: :text_field, step: :any
+    = f.number_field :amount, class: :text_field, step: :any
```
Comment (fiestacasey) [inline app/views/admin/products/_form.html.haml]: step of 0.01 would make more sense. You wouldn't charge someone 7.0004323 for something

**Review by fiestacasey — CHANGES_REQUESTED** [review]

**Comment by fiestacasey** [comment]
@ksolkowski @mikeSheets @austinkahly

File: app/models/coupon.rb [hunk 5]
```diff
@@ -3,7 +3,11 @@ class Coupon < ApplicationRecord
...
   validates :amount, :name, :coupon_type, presence: true
   has_many :order_items, as: :source, dependent: :destroy
+
+  def negate_amount
+    self.amount *= -1
```
Comment (ksolkowski) [inline app/models/coupon.rb]: ```-self.amount```

File: app/assets/javascripts/Resources/coupon.coffee [hunk 6]
```diff
@@ -0,0 +1,13 @@
+app = angular.module('customerApp')
+
+app.factory('Coupon', ['$resource', ($resource) ->
```
Comment (ksolkowski) [inline app/assets/javascripts/Resources/coupon.coffee:3]: why is the folder name ```Resources```?

File: config/routes.rb [hunk 7]
```diff
@@ -22,6 +22,9 @@
       resources :credit_cards do
         put :update
       end
+      resources :coupons do
+        put :update
```
Comment (ksolkowski) [inline config/routes.rb]: ```put :update``` should already be set up with ```resources```

**Review by ksolkowski — CHANGES_REQUESTED** [review]

File: app/views/admin/products/_form.html.haml (hunk 4 above)
Comment (gsnovak) [reply app/views/admin/products/_form.html.haml]: @fiestacasey how can i actually set this? form_for won't allow me to enter in "0.01" or 0.01 for the step.

File: app/assets/javascripts/Resources/coupon.coffee (hunk 6 above)
Comment (fiestacasey) [reply app/assets/javascripts/Resources/coupon.coffee:3]: cause @gsnovak just can't let go of his .net ways

File: app/views/admin/products/_form.html.haml (hunk 4 above)
Comment (gsnovak) [reply app/views/admin/products/_form.html.haml]: @fiestacasey i've concluded that this is impossible and that i don't like form_for 👍

Comment (austinkahly) [reply app/views/admin/products/_form.html.haml]: @gsnovak Did you try removing the class from this? 
`= f.number_field :amount, step: 0.01`?

File: app/assets/javascripts/controllers/checkout_controller.coffee [hunk 8]
```diff
@@ -79,6 +80,22 @@ app.controller 'CheckoutController',['$window', '$q', '$scope', 'Order', 'Addres
...
         $scope.savingAddress = false
 
+   $scope.applyCoupon = (couponCode) ->
+    return if $scope.grabbingCoupon
+    $scope.grabbingCoupon = true
+    Coupon.query(code: couponCode).$promise
+      .then (data) ->
+        coupon = new Coupon(data)
+        itemToSave = new OrderItem(quantity: 1, source_type: "Coupon", source_id: coupon.id, order_id: $scope.cartOrder.id )
+        itemToSave.$save().then ->
```
Comment (austinkahly) [inline app/assets/javascripts/controllers/checkout_controller.coffee]: do you want to do `itemToSave.$save().then (item) ->`? This way you get the updated item with its id and what not.

File: app/views/checkout/index.html.haml [hunk 9]
```diff
@@ -118,6 +118,16 @@
               %td{ colspan: "4"}
                 %strong
                   Total: {{ total() | currency }}
+        .alert.alert-danger{ ng: { if: 'couponErrors' }}
+          %ul{ ng: { repeat: "val in couponErrors" }}
+            %li
+              {{ val }}
+        .col-sm-3
+          %label.control-label{ for: "coupon"} Coupon
```
Comment (austinkahly) [inline app/views/checkout/index.html.haml]: `{ for: "coupon" }`

File: app/views/checkout/index.html.haml [hunk 10]
```diff
@@ -118,6 +118,16 @@
...
+        .col-sm-3
+          %input#coupon{ type: "text", class: "form-control", ng: { model: "couponCode"} }
```
Comment (austinkahly) [inline app/views/checkout/index.html.haml]: could this be `%input.form-control#coupon`?

File: db/migrate/20170814213115_add_amount_to_products.rb [hunk 11]
```diff
@@ -0,0 +1,5 @@
+class AddAmountToProducts < ActiveRecord::Migration[5.1]
+  def change
+    add_column :products, :amount, :float
```
Comment (austinkahly) [inline db/migrate/20170814213115_add_amount_to_products.rb:3]: you could have done a `rename_column :products, :price, :amount`

Comment (gsnovak) [reply db/migrate/20170814213115_add_amount_to_products.rb:3]: didn't know that was a thing. cool!

File: app/views/admin/products/_form.html.haml (hunk 4 above)
Comment (gsnovak) [reply app/views/admin/products/_form.html.haml]: i figured out what i was doing wrong. lets never speak of this again.

File: app/assets/javascripts/controllers/checkout_controller.coffee (hunk 8 above)
Comment (gsnovak) [reply app/assets/javascripts/controllers/checkout_controller.coffee]: stop making so much sense

File: app/views/admin/coupons/_form.html.haml [hunk 12]
```diff
@@ -2,6 +2,6 @@
   = bootstrap_form_for [:admin, @source_model] do |f|
     = f.text_field :code
     = f.text_field :name
-    = f.number_field :amount
-    = f.text_field :coupon_type  
+    = f.number_field :amount, step: 0.01
```
Comment (austinkahly) [inline app/views/admin/coupons/_form.html.haml:5]: HEY MAN YOU GOT IT!

Comment (fiestacasey) [reply app/views/admin/coupons/_form.html.haml:5]: i never doubted you.

File: app/views/checkout/index.html.haml (hunk 9 above)
Comment (gsnovak) [reply app/views/checkout/index.html.haml]: what?

Comment (austinkahly) [reply app/views/checkout/index.html.haml]: add a space after coupon.

Comment (gsnovak) [reply app/views/checkout/index.html.haml]: huh?

File: app/views/checkout/index.html.haml [hunk 13]
```diff
@@ -118,6 +118,16 @@
...
                 %strong
                   Total: {{ total() | currency }}
+        .alert.alert-danger{ ng: { if: 'couponErrors' }}
+          %ul{ ng: { repeat: "val in couponErrors" }}
+            %li
+              {{ val }}
+        .col-sm-3
+          %label.control-label{ for: "coupon" } Coupon
+        .col-sm-3
+          %input.form-control#coupon{ type: "text", class: "form-control", ng: { model: "couponCode"} }
```
Comment (austinkahly) [inline app/views/checkout/index.html.haml]: hey silly, you can remove the class part now.

**Comment by gsnovak** [comment]
@fiestacasey @austinkahly lets take this thing out back and put it out of its misery

File: app/views/checkout/index.html.haml [hunk 14]
```diff
@@ -118,6 +118,16 @@
...
                 %strong
                   Total: {{ total() | currency }}
+        .alert.alert-danger{ ng: { if: 'couponErrors' }}
+          %ul{ ng: { repeat: "val in couponErrors" }}
+            %li
+              {{ val }}
+        .col-sm-3
+          %label.control-label Coupon
+        .col-sm-3
+          %input.form-control{ type: "text", class: "form-control", ng: { model: "couponCode"} }
```
Comment (austinkahly) [inline app/views/checkout/index.html.haml]: @gsnovak Sure, we can do that after you remove the redundant `class: "form-control"`

Comment (gsnovak) [reply app/views/checkout/index.html.haml]: whoops

**Comment by fiestacasey** [comment]
good to go on this. this project is now complete

**Pull request merged**